src/System.Private.ServiceModel/src/System/ServiceModel/Channels/SecurityChannelFactory.cs

```diff
@@ -0,0 +1,843 @@
+        {
+            get
+            {
+                ThrowIfProtocolFactoryNotSet();
+                return _securityProtocolFactory.SupportsRequestReply;
+            }
+        }
+
+        public MessageVersion MessageVersion
+        {
+            get
+            {
+                return _messageVersion;
+            }
+        }
+
+        void CloseProtocolFactory(bool aborted, TimeSpan timeout)
+        {
+            if (_securityProtocolFactory != null && !SessionMode)
+            {
+                _securityProtocolFactory.Close(aborted, timeout);
+                _securityProtocolFactory = null;
+            }
+        }
+
+        async Task CloseProtocolFactoryAsync(TimeSpan timeout)
+        {
+            if (_securityProtocolFactory != null && !SessionMode)
+            {
+                await _securityProtocolFactory.CloseAsync(timeout);
+                _securityProtocolFactory = null;
+            }
+        }
+
+        public override T GetProperty<T>()
+        {
+            if (SessionMode && (typeof(T) == typeof(IChannelSecureConversationSessionSettings)))
+            {
+                return (T)(object)SessionClientSettings;
+            }
+            else if (typeof(T) == typeof(ISecurityCapabilities))
+            {
+                return (T)(object)_securityCapabilities;
+            }
+
+            return base.GetProperty<T>();
+        }
+
+        protected override void OnAbort()
+        {
+            base.OnAbort();
+            CloseProtocolFactory(true, TimeSpan.Zero);
+            if (_sessionClientSettings != null)
+            {
+                _sessionClientSettings.Abort();
+            }
+        }
+
+        protected override IAsyncResult OnBeginClose(TimeSpan timeout, AsyncCallback callback, object state)
+        {
+            throw ExceptionHelper.PlatformNotSupported("SecurityChannelFactory OnBeginClose not supported.");
+
+            //List<OperationWithTimeoutBeginCallback> begins = new List<OperationWithTimeoutBeginCallback>();
+            //List<OperationEndCallback> ends = new List<OperationEndCallback>();
+            //begins.Add(new OperationWithTimeoutBeginCallback(base.OnBeginClose));
+            //ends.Add(new OperationEndCallback(base.OnEndClose));
+
+            //if (securityProtocolFactory != null && !SessionMode)
+            //{
+            //    begins.Add(new OperationWithTimeoutBeginCallback(securityProtocolFactory.BeginClose));
+            //    ends.Add(new OperationEndCallback(securityProtocolFactory.EndClose));
+            //}
+
+            //if (sessionClientSettings != null)
+            //{
+            //    begins.Add(new OperationWithTimeoutBeginCallback(sessionClientSettings.BeginClose));
+            //    ends.Add(new OperationEndCallback(sessionClientSettings.EndClose));
+            //}
+
+            //return OperationWithTimeoutComposer.BeginComposeAsyncOperations(timeout, begins.ToArray(), ends.ToArray(), callback, state);
```

Comment:
These comments are going to be deleted as we wire up the Task async code, but I'd prefer to leave them as reference code till then. 

---
In reply to: [78341973](https://github.com/dotnet/wcf/pull/1515#discussion_r78341973) [](ancestors = 78341973)